### PR TITLE
ddl: apply common handle to index reorganization

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [Clustered index](https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-cluster-index.md) 

What's Changed:
When adding an index to a table, a data reorganization procedure is launched to backfilling index data. The original range and progress marker is `int64`, now it is changed to `kv.Handle`.

How it Works:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Consumes more MEM

### Release note 

- N/A

### CODE REVIEW

File: meta/meta_test.go [hunk 1]
```diff
@@ -327,36 +339,42 @@ func (s *testSuite) TestDDL(c *C) {
 	err = t.UpdateDDLJob(0, job, true)
 	c.Assert(err, IsNil)
 
-	err = t.UpdateDDLReorgStartHandle(job, 1)
+	// There are 3 meta key relate to index reorganization:
+	// start_handle, end_handle and physical_table_id.
+	// Only start_handle is initialized.
+	err = t.UpdateDDLReorgStartHandle(job, kv.IntHandle(1))
```
Comment (coocood) [inline meta/meta_test.go:345]: Why not use `newHandle(1)`?

File: ddl/index.go [hunk 2]
```diff
@@ -1072,6 +1072,18 @@ func (w *addIndexWorker) batchCheckUniqueKey(txn kv.Transaction, idxRecords []*i
 	return nil
 }
 
+func encodeHandleInUniqueIndexValueForBatchCheck(h kv.Handle) []byte {
+	if h.IsInt() {
+		return tables.EncodeHandleInUniqueIndexValue(h.IntValue())
+	}
+	handleLen := h.Len()
+	data := make([]byte, 4, 1+1+2+handleLen)
+	data[2] = uint8(handleLen >> 8)
```
Comment (coocood) [inline ddl/index.go:1081]: We should have a single encode common handle method.

File: ddl/index.go [hunk 3]
```diff
@@ -1072,6 +1072,18 @@ func (w *addIndexWorker) batchCheckUniqueKey(txn kv.Transaction, idxRecords []*i
 	return nil
 }
 
+func encodeHandleInUniqueIndexValueForBatchCheck(h kv.Handle) []byte {
```
Comment (coocood) [inline ddl/index.go:1075]: We can modify `tables.EncodeHandleInUniqueIndexValue` to support common handle and use it there https://github.com/pingcap/tidb/blob/master/table/tables/index.go#L373

File: meta/meta_test.go (hunk 1 above)
Comment (tangenta) [reply meta/meta_test.go:345]: This is related to the test below. Old-version TiDB does not set `physical_table_id`, and it always returns `maxInt64` as `end_handle`, so the type of these handles should be `Int`.

File: table/tables/index.go [hunk 4]
```diff
@@ -37,10 +37,13 @@ import (
...
-	var data [8]byte
-	binary.BigEndian.PutUint64(data[:], uint64(h))
-	return data[:]
+func EncodeHandleInUniqueIndexValue(h kv.Handle) []byte {
+	if h.IsInt() {
+		var data [8]byte
+		binary.BigEndian.PutUint64(data[:], uint64(h.IntValue()))
+		return data[:]
+	}
+	return encodeCommonHandle(nil, h)
```
Comment (coocood) [inline table/tables/index.go:46]: encodeCommonHandle([]byte{0}, h)

File: table/tables/index.go [hunk 5]
```diff
@@ -394,7 +397,7 @@ func (c *index) Create(sctx sessionctx.Context, rm kv.RetrieverMutator, indexedV
 		idxVal = make([]byte, 0)
 		if distinct {
 			if h.IsInt() {
-				idxVal = EncodeHandleInUniqueIndexValue(h.IntValue())
+				idxVal = EncodeHandleInUniqueIndexValue(h)
```
Comment (coocood) [inline table/tables/index.go:400]: If we add `untouched` parameter to `EncodeHandleInUniqueIndexValue`, we can call it out of the `if h.IsInt` block

**Comment by coocood** [comment]
LGTM

File: util/admin/admin.go [hunk 6]
```diff
@@ -77,7 +77,11 @@ func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
 		return info, nil
 	}
 
-	info.ReorgHandle, _, _, err = t.GetDDLReorgHandle(addIdxJob)
+	tbl, err := t.GetTable(job.SchemaID, job.TableID)
```
Comment (coocood) [inline util/admin/admin.go:80]: Paniced here, job is nil

**Pull request merged**